### PR TITLE
Enable the Scorecard Github Action and badge

### LINKED ISSUE #926: Enable the Scorecard Github Action and badge
Hi I am Joyce and I'm working on behalf of Google and the [OpenSSF][ossf] to help essential open-source projects improve their supply-chain security. The OpenSSF is a non-profit foundation backed by the Linux Foundation, dedicated to improving the security of the open-source community. It counts GitHub as a founding [member][ossf-membership].

**What would you like to be added**:
I would like to suggest for you to use Scorecard Github Action to increase the security of your repository.

The [Scorecard][sc] system combines dozens of automated checks to let maintainers better understand their project's supply-chain security posture. It is developed by the OpenSSF, with [direct support from GitHub][sc-blog].

However, the OpenSSF has also developed the [Scorecard GitHub Action][sc-gha], which adds the results of its checks to the project's [security dashboard](https://github.com/anchore/grype/security), as well as suggestions on how to solve any issues (see examples on the "Additional context"). This Action has been adopted by 1600+ projects already.

Would you be interested in a PR which adds this Action? Optionally, it can also publish your results to the OpenSSF REST API, which allows a [badge][badge] with the project's score to be added to its README.

**Why is this needed**:

Considering the relevance grype has on ensuring the security of many projects, it's been included in the OpenSSF's list of the 100 most critical open-source projects. Thus, the Scorecard Github Action adoption in this project would help you on the identification and the solution of security risks of the repository, which would increase the security of the project and garantee that it is (mostly) safe from malicious sabotage.

**Additional context**:

![Code scanning dashboard with multiple alerts, including Code-Review and Token-Permissions][img-security]

![Detail of a Token-Permissions alert, indicating the specific file and remediation steps][img-detail]

[badge]: https://openssf.org/blog/2022/09/08/show-off-your-security-score-announcing-scorecards-badges/
[ossf]: https://openssf.org/
[ossf-membership]: https://openssf.org/about/members/
[sc]: https://github.com/ossf/scorecard
[sc-blog]: https://github.blog/2022-01-19-reducing-security-risk-oss-actions-opensff-scorecards-v4/
[sc-gha]: https://github.com/ossf/scorecard-action
[img-security]: https://user-images.githubusercontent.com/15221358/190184391-84ca1844-259a-4b3b-9c86-74adadbea7f1.png
[img-detail]: https://user-images.githubusercontent.com/15221358/190184600-ee8d3b39-077e-416a-8711-1b5fb01cf0b3.png

### PR DESCRIPTION
**Pull request by joycebrum**
Closes #926 

Hi, here is the PR with the changes to enable the OpenSSF Scorecard Gihub Action as described at #926 

Thanks for the attention and any doubts or concerns please reach me out.

### CODE REVIEW

File: .github/workflows/scorecards.yml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+          publish_results: true
+
+      # Upload the results as artifacts (optional). Commenting out will disable uploads of run results in SARIF
+      # format to the repository Actions tab.
+      - name: "Upload artifact"
+        uses: actions/upload-artifact@6673cd052c4cd6fcf4b4e6e60ea986c889389535 # tag=v3.0.0
+        with:
+          name: SARIF file
+          path: results.sarif
+          retention-days: 5
```
Comment (kzantow) [inline .github/workflows/scorecards.yml:48]: We probably don't need this

File: .github/workflows/scorecards.yml [hunk 2]
```diff
@@ -0,0 +1,54 @@
+name: Scorecards supply-chain security
+on:
+  # Only the default branch is supported.
+  branch_protection_rule:
+  schedule:
+    - cron: '34 10 * * 2'
```
Comment (kzantow) [inline .github/workflows/scorecards.yml:6]: Is there a need for anything other than a push to `main`?

**Review by kzantow — COMMENTED** [review]
Thanks for the contribution! One comment: we looked at doing this before but have a number of binaries and similar artifacts committed to the repository for testing, is there a way to explicitly omit such things from scorecard?

File: .github/workflows/scorecards.yml (hunk 2 above)
Comment (joycebrum) [reply .github/workflows/scorecards.yml]: It is not necessary, it's just two different trigger types. Even if there was no updates on the main branch, the Scorecard Action would still weekly search for vulnerabilities.

I've removed the schedule trigger but if you prefer to keep it let me know that I'll add it again

File: .github/workflows/scorecards.yml (hunk 1 above)
Comment (joycebrum) [reply .github/workflows/scorecards.yml]: Yeah, its fully optional, I've removed it

**Comment by joycebrum** [comment]
> Thanks for the contribution! One comment: we looked at doing this before but have a number of binaries and similar artifacts committed to the repository for testing, is there a way to explicitly omit such things from scorecard?

About this topic, I've looked it into the details to bring a more correct response and I seems that it is not possible to configure a "ignore list" yet, but it is a issue that the Scorecard team is working on (https://github.com/ossf/scorecard/issues/1270).

Besides, I would like to bring to attention that only one check really look to binarie files, which is the Binary-Artifacts check. This check just points out if there is a binary file in the repo or not. All the other checks do not look or scan through binary files.

**Comment by westonsteimel** [comment]
@kzantow I don't think the binary one hurts the score very much anyways - looked like it was 9 to me for that one.  The one that brings it down the most appears to be not pinning GitHub Actions to a hash

**Comment by kzantow** [comment]
Hi @joycebrum -- sorry for the delay getting back to this. I've had a look through the scorecard results and I'm trying to understand a few things. First of all, the results I see are:

**Overall score: 6.4**

Based on what I found in the JSON:

| Check | Score |
| --- | ---
| Maintained | 10 | 
| Code review | 10
| CII best practices | \<missing>
| Vulnerabilities | 10
| Packaging | -1
| License | 10
| Dangerous workflow | 10
| Token permissions | \<missing>
| Binary artifacts | 9
| Security policy | \<missing>
| Signed releases | \<missing>
| Branch protection | 8
| Dependency update tool | 10
| Fuzzing | <missing>
| Pinned dependencies | 5
| SAST | -1

I'd like to understand how this score is calculated. I assume missing values are zero? What about `-1` values -- are these unknown and omitted? (I just found the section explaining the weighting, so haven't done the math with weighted numbers yet...)

Some of these seem to be misleading, for example we are running codeql already (albeit only on pushes to `main` due to the time the check takes): https://github.com/anchore/syft/blob/main/.github/workflows/codeql-analysis.yml

And we are signing at least some of the release artifacts (e.g. macOS builds).

Is there a way to exclude specific checks because they aren't applicable or giving incorrect results?

**Comment by joycebrum** [comment]
Hi @kzantow, don't worry about the time at all.

Running the scorecard locally here the results I've got are a little more human friendly, so it might help us better understand the results:

| SCORE | NAME                   | REASON                                                                                                                                                                                                                                                                                                                 |
|-------|------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| 9/10  | Binary-Artifacts       | There is at least one binary artifact in the source code                                                                                                                                                                                                                                                               |
| 8/10  | Branch-Protection      | It is 0, 3, 6, 8, 9 or 10 according to https://github.com/ossf/scorecard/blob/main/docs/checks.md#branch-protection                                                                                                                                                                                                    |
| 10/10 | CI-Tests               |                                                                                                                                                                                                                                                                                                                        |
| 0/10  | CII-Best-Practices     | The <missing> here really means 0, because the evaluation of this check depends onwhether the project has earned an [OpenSSF (formerly CII) Best Practices Badge](https://bestpractices.coreinfrastructure.org/)                                                                                                       |
| 10/10 | Code-Review            |                                                                                                                                                                                                                                                                                                                        |
| 10/10 | Contributors           |                                                                                                                                                                                                                                                                                                                        |
| 10/10 | Dangerous-Workflow     |                                                                                                                                                                                                                                                                                                                        |
| 10/10 | Dependency-Update-Tool |                                                                                                                                                                                                                                                                                                                        |
| 0/10  | Fuzzing                | Project is not Fuzzed                                                                                                                                                                                                                                                                                                  |
| 10/10 | License                |                                                                                                                                                                                                                                                                                                                        |
| 10/10 | Maintained             |                                                                                                                                                                                                                                                                                                                        |
| ?     | Packaging              | "no published package detected". The <missing> here doesn't mean 0, it means that the check is not "applicable" so it will not be counted in the score                                                                                                                                                                 |
| 5/10  | Pinned Dependencies    | dependency not pinned by hash detected -- score normalized to 5                                                                                                                                                                                                                                                        |
| 7/10  | SAST                   | "SAST tool detected but do not run on all commits"                                                                                                                                                                                                                                                                      |
| 0/10  | Security-Policy        | "Security Policy file not detected" - In this case, it is a 0 or 10 check, but it is also a quick-win                                                                                                                                                                                                                  |
| 0/10  | Signed-Releases        | "0 out of 5 artifacts are signed or have provenance" - It seemed to find 5 artifacts in the repo that were not signed.                                                                                                                                                                                                 |
| 0/10  | Token-Permissions      | "non read-only tokens detected in Github Workflow" - According to [Token-Permissions Check](https://github.com/ossf/scorecard/blob/2cbf5afd5460b51fd40939f8c44b32543b1a0bcb/docs/checks.md#token-permissions), this check determines whether the project's automated workflows tokens are set to read-only by default. |
| 10/10 | Vulnerabilities        |                                                                                                                                                                                                                                                                                                                        |

About the score calculation there is the weights:

- “Critical” risk checks are weighted at 10
- “High” risk checks are weighted at 7.5
- “Medium” risk checks are weighted at 5
- “Low” risk checks are weighted at 2.5

This we would have:

( 10 * (Dangerous-Workflow) + 7.5 * (Binary-Artifacts + Branch-Protection + Code-Review + Dependency-Update-Tool + Maintained + Signed-Releases + Token-Permissions + Vulnerabilities) + 5 * (Fuzzing + Pinned-Dependencies + <s>Packaging</s> + SAST + Security-Policy) + 2.5 * (CI-Tests + CII-Best-Practices + Contributors + License) ) / (10 + 60 + 20 + 10)

(100 + 427.5 + 60 + 75) / 100 = 6,625 (which was the value that the scorecard calculate in the CLI)

![image](https://user-images.githubusercontent.com/22223372/198717801-8cc803d6-f785-4e98-aa7c-8f22878525ff.png)

**Comment by westonsteimel** [comment]
And it looks like the signed-releases and token-permissions would boost the score the most?

**Comment by kzantow** [comment]
Thanks for the contribution @joycebrum -- we're going to move forward with this as-is and work to make incremental improvements 👍

**Comment by joycebrum** [comment]
Great to hear that @kzantow. I would like to also offer my help on working in any issue from scorecard checks (or any other security issue tbh) you might want to implement. Feel free to ping me and/or assign it to me 😄

**Review by spiffcs — APPROVED** [review]

**Pull request merged**